File: rk3328/gpiotest/client_gpio.py
#!/usr/bin/python3

import socket
import time
import sys
from PyQt5.QtWidgets import (QWidget, QMainWindow, QGridLayout, QPushButton, QApplication, QLineEdit)

HOST = '192.168.86.33'   # The remote host
PORT = 50007              # The same port as used by the server

sock=socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.connect((HOST, PORT))

class Client_gpio(QMainWindow):

    def __init__(self):
        super().__init__()

        self.initUI()


    def initUI(self):

        self.widget = QWidget()
        self.setCentralWidget(self.widget)

        StartTestbtn = QPushButton("Start Test", self)
        StopTestbtn = QPushButton("Stop Test", self)
        LedOnbtn = QPushButton("Led On", self)
        LedOffbtn = QPushButton("Led Off", self)
        Lcdredbtn = QPushButton("Lcd Red", self)
        Lcdgreenbtn = QPushButton("Lcd Green", self)
        Lcdbluebtn = QPushButton("Lcd Blue", self)
        Lcdyellowbtn = QPushButton("Lcd Yellow", self)
        Lcdcyanbtn = QPushButton("Lcd Cyan", self)
        Lcdpurplebtn = QPushButton("Lcd Purple", self)
        Lcdwhitebtn = QPushButton("Lcd White", self)
        Lcdblackbtn = QPushButton("Lcd Black", self)
        self.qle = QLineEdit(self)
        Lcdsendtxtbtn = QPushButton("Lcd Sendtxt", self)
        Servoleftbtn = QPushButton("Servo Left", self)
        Servomiddlebtn = QPushButton("Servo Middle", self)
        Servorightbtn = QPushButton("Servo Right", self)
        Servostopbtn = QPushButton("Servo Stop", self)

        grid = QGridLayout(self.widget)
        grid.setSpacing(10)
        grid.addWidget(StartTestbtn, 0, 0)
        grid.addWidget(StopTestbtn, 0, 3)
        grid.addWidget(LedOnbtn, 1, 0)
        grid.addWidget(LedOffbtn, 1, 1)
        grid.addWidget(Lcdredbtn, 2, 0)
        grid.addWidget(Lcdgreenbtn, 2, 1)
        grid.addWidget(Lcdbluebtn, 2, 2)
        grid.addWidget(Lcdyellowbtn, 2, 3)
        grid.addWidget(Lcdcyanbtn, 3, 0)
        grid.addWidget(Lcdpurplebtn, 3, 1)
        grid.addWidget(Lcdwhitebtn, 3, 2)
        grid.addWidget(Lcdblackbtn, 3, 3)
        grid.addWidget(self.qle, 3, 0, 3, 2)
        grid.addWidget(Lcdsendtxtbtn, 4, 3)
        grid.addWidget(Servoleftbtn, 5, 0)
        grid.addWidget(Servomiddlebtn, 5, 1)
        grid.addWidget(Servorightbtn, 5, 2)
        grid.addWidget(Servostopbtn, 5, 3)

        self.widget.setLayout(grid)

        StartTestbtn.clicked.connect(self.buttonClicked)
        StopTestbtn.clicked.connect(self.buttonClicked)
        LedOnbtn.clicked.connect(self.buttonClicked)
        LedOffbtn.clicked.connect(self.buttonClicked)
        Lcdredbtn.clicked.connect(self.buttonClicked)
        Lcdgreenbtn.clicked.connect(self.buttonClicked)
        Lcdbluebtn.clicked.connect(self.buttonClicked)
        Lcdyellowbtn.clicked.connect(self.buttonClicked)
        Lcdcyanbtn.clicked.connect(self.buttonClicked)
        Lcdpurplebtn.clicked.connect(self.buttonClicked)
        Lcdwhitebtn.clicked.connect(self.buttonClicked)
        Lcdblackbtn.clicked.connect(self.buttonClicked)
        self.qle.textChanged[str].connect(self.onChanged)
        Lcdsendtxtbtn.clicked.connect(self.buttonClicked)
        Servoleftbtn.clicked.connect(self.buttonClicked)
        Servomiddlebtn.clicked.connect(self.buttonClicked)
        Servorightbtn.clicked.connect(self.buttonClicked)
        Servostopbtn.clicked.connect(self.buttonClicked)

        self.statusBar()

        self.setGeometry(300, 300, 400, 300)
        self.setWindowTitle('gpio test')
        self.show()

    def onChanged(self, text):

        self.qle.setText(text)
        self.update()

    def buttonClicked(self):

        sender = self.sender()
        if sender.text() == "Start Test":
            sock.sendall(b"start")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Stop Test":
            sock.sendall(b"bye")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Led On":
            sock.sendall(b"led_on")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Led Off":
            sock.sendall(b"led_off")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        if sender.text() == "Lcd Red":
            sock.sendall(b"lcd_red")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Lcd Green":
            sock.sendall(b"lcd_green")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Lcd Blue":
            sock.sendall(b"lcd_blue")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Lcd Yellow":
            sock.sendall(b"lcd_yellow")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Lcd Cyan":
            sock.sendall(b"lcd_cyan")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Lcd Purple":
            sock.sendall(b"lcd_purple")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Lcd White":
            sock.sendall(b"lcd_white")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Lcd Black":
            sock.sendall(b"lcd_black")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Lcd Sendtxt":
            if self.qle.text():
                msg = self.qle.text()
                msg = "Hello" + msg
                sock.sendall(bytearray(msg.encode('utf-8')))
                data = sock.recv(1024)
                self.statusBar().showMessage(repr(data))
        elif sender.text() == "Servo Left":
            sock.sendall(b"servo_left")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Servo Middle":
            sock.sendall(b"servo_middle")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Servo Right":
            sock.sendall(b"servo_right")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))
        elif sender.text() == "Servo Stop":
            sock.sendall(b"servo_stop")
            data = sock.recv(1024)
            self.statusBar().showMessage(repr(data))


if __name__ == '__main__':

    app = QApplication(sys.argv)
    ex = Client_gpio()
    sys.exit(app.exec_())

File: radxa/gpiotest/client_gpio.cpp
#include "client_gpio.h"
#include <QGridLayout>
#include <QLabel>
#include <QStatusBar>
#include <QLineEdit>
#include <stdio.h> 
#include <sys/socket.h> 
#include <arpa/inet.h>
#include <netinet/in.h> 
#include <netdb.h>
#include <unistd.h>
#include <time.h>
#include <string.h> 
#include <stdlib.h>

#define HOST "192.168.86.34"
#define PORT 8888

int make_socket( uint16_t port )
{
    int sock;
    struct sockaddr_in serv_addr;

    if ((sock = socket(AF_INET, SOCK_STREAM, 0)) < 0)
    {
        printf("\n Socket creation error \n");
        return -1;
    }

    memset(&serv_addr, '0', sizeof(serv_addr));

    serv_addr.sin_family = AF_INET;
    serv_addr.sin_port = htons( port );

    // Convert IPv4 and IPv6 addresses from text to binary form
    if(inet_pton(AF_INET, HOST, &serv_addr.sin_addr)<=0)
    {
        printf("\nInvalid address/ Address not supported \n");
        return -1;
    }

    if (connect(sock, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0)
    {
        printf("\nConnection Failed \n");
        return -1;
    }
    return sock;
}

Client_gpio::Client_gpio(QWidget *parent)
    : QMainWindow(parent) {

  QWidget* widget = new QWidget(this);

  QPushButton *StartTestBtn = new QPushButton("Start Test", this);
  QPushButton *StopTestBtn = new QPushButton("Stop Test", this);
  QPushButton *ledOnBtn = new QPushButton("Led On", this);
  QPushButton *ledOffBtn = new QPushButton("Led Off", this);
  QPushButton *redBtn = new QPushButton("Lcd Red", this);
  QPushButton *greenBtn = new QPushButton("Lcd Green", this);
  QPushButton *blueBtn = new QPushButton("Lcd Blue", this);
  QPushButton *yellowBtn = new QPushButton("Lcd Yellow", this);
  QPushButton *cyanBtn = new QPushButton("Lcd Cyan", this);
  QPushButton *purpleBtn = new QPushButton("Lcd Purple", this);
  QPushButton *whiteBtn = new QPushButton("Lcd White", this);
  QPushButton *blackBtn = new QPushButton("Lcd Black", this);
  QPushButton *textBtn = new QPushButton("send text", this);
  QPushButton *servoRightBtn = new QPushButton("Servo Right", this);
  QPushButton *servoLeftBtn = new QPushButton("Servo Left", this);
  QPushButton *servoMiddleBtn = new QPushButton("Servo Middle", this);
  QPushButton *servoStopBtn = new QPushButton("Servo Stop", this);
  labelR = new QLabel("R", this);
  labelR->setStyleSheet("QLabel{color:rgba(255, 0, 0, 255);}");
  labelG = new QLabel("G", this);
  labelG->setStyleSheet("QLabel{color:rgba(0, 255, 0, 255);}");
  labelB = new QLabel("B", this);
  labelB->setStyleSheet("QLabel{color:rgba(0, 0, 255, 255);}");
  sliderR = new QSlider(Qt::Horizontal, this);
  sliderR->setRange(0, 255);
  sliderG = new QSlider(Qt::Horizontal, this);
  sliderG->setRange(0, 255);
  sliderB = new QSlider(Qt::Horizontal, this);
  sliderB->setRange(0, 255);
  labelServo = new QLabel("Servo", this);
  labelServo->setStyleSheet("QLabel{color:rgba(125, 64, 255, 255);}");
  sliderServo = new QSlider(Qt::Horizontal, this);
  sliderServo->setRange(50, 100);

  QGridLayout *grid = new QGridLayout(widget);
  grid->setSpacing(10);
  grid->addWidget(StartTestBtn, 0, 0);
  grid->addWidget(StopTestBtn, 0, 3);
  grid->addWidget(ledOnBtn, 1, 0);
  grid->addWidget(ledOffBtn, 1, 1);
  grid->addWidget(redBtn, 2, 0);
  grid->addWidget(greenBtn, 2, 1);
  grid->addWidget(blueBtn, 2, 2);
  grid->addWidget(yellowBtn, 2, 3);
  grid->addWidget(cyanBtn, 3, 0);
  grid->addWidget(purpleBtn, 3, 1);
  grid->addWidget(whiteBtn, 3, 2);
  grid->addWidget(blackBtn, 3, 3);

  QLineEdit *edt2 = new QLineEdit(widget);
  grid->addWidget(edt2, 3, 0, 3, 3);
  grid->addWidget(textBtn, 4, 3);
  grid->addWidget(servoRightBtn, 5, 2);
  grid->addWidget(servoLeftBtn, 5, 0);
  grid->addWidget(servoMiddleBtn, 5, 1);
  grid->addWidget(servoStopBtn, 5, 3);
  grid->addWidget(sliderR, 5, 0, 5, 3);
  grid->addWidget(labelR, 7, 3);
  grid->addWidget(sliderG, 6, 0, 6, 3);
  grid->addWidget(labelG, 8, 3);
  grid->addWidget(sliderB, 7, 0, 7, 3);
  grid->addWidget(labelB, 10, 3);
  grid->addWidget(sliderServo, 8, 0, 8, 3);
  grid->addWidget(labelServo, 11, 3);

  widget->setLayout(grid);
  setCentralWidget(widget);
  statusBar();

  connect(StartTestBtn, &QPushButton::clicked, this, &Client_gpio::start_test);
  connect(StopTestBtn, &QPushButton::clicked, this, &Client_gpio::stop_test);
  connect(ledOnBtn, &QPushButton::clicked, this, &Client_gpio::ledOn);
  connect(ledOffBtn, &QPushButton::clicked, this, &Client_gpio::ledOff);
  connect(redBtn, &QPushButton::clicked, this, &Client_gpio::OnRed);
  connect(greenBtn, &QPushButton::clicked, this, &Client_gpio::OnGreen);
  connect(blueBtn, &QPushButton::clicked, this, &Client_gpio::OnBlue);
  connect(yellowBtn, &QPushButton::clicked, this, &Client_gpio::OnYellow);
  connect(cyanBtn, &QPushButton::clicked, this, &Client_gpio::OnCyan);
  connect(purpleBtn, &QPushButton::clicked, this, &Client_gpio::OnPurple);
  connect(whiteBtn, &QPushButton::clicked, this, &Client_gpio::OnWhite);
  connect(blackBtn, &QPushButton::clicked, this, &Client_gpio::OnBlack);
  connect(edt2, &QLineEdit::textChanged, this, &Client_gpio::setText);
  connect(textBtn, &QPushButton::clicked, this, &Client_gpio::sendText);
  connect(servoRightBtn, &QPushButton::clicked, this, &Client_gpio::servoRight);
  connect(servoLeftBtn, &QPushButton::clicked, this, &Client_gpio::servoLeft);
  connect(servoMiddleBtn, &QPushButton::clicked, this, &Client_gpio::servoMiddle);
  connect(servoStopBtn, &QPushButton::clicked, this, &Client_gpio::servoStop);
  connect(sliderR, &QSlider::valueChanged, this, &Client_gpio::send_lcd);
  connect(sliderG, &QSlider::valueChanged, this, &Client_gpio::send_lcd);
  connect(sliderB, &QSlider::valueChanged, this, &Client_gpio::send_lcd);
  connect(sliderServo, &QSlider::valueChanged, this, &Client_gpio::send_servo);

  m_text = QString();
  sock = make_socket( PORT );
}

void Client_gpio::send_lcd() {
  int valread;
  char buffer[1024] = {0};
  char lcd[10];
  sprintf(lcd, "%03d%03d%03d", sliderR->value(), sliderG->value(), sliderB->value());
  send(sock , lcd, strlen(lcd) , 0 );
  valread = read( sock, buffer, 1024 );
  buffer[valread] = '\0';
  statusBar()->showMessage(buffer);
}

void Client_gpio::send_servo() {

  int valread;
  char servo[4];
  sprintf(servo, "%d", sliderServo->value() + 300);
  char buffer[1024] = {0};
  send(sock , servo, strlen(servo) , 0 );
  valread = read( sock, buffer, 1024 );
  buffer[valread] = '\0';
  statusBar()->showMessage(buffer);
}
   
void Client_gpio::send_cmd( const char * cmd ) {

  int valread;
  char buffer[1024] = {0};
  send(sock , cmd, strlen(cmd) , 0 );
  valread = read( sock, buffer, 1024 );
  buffer[valread] = '\0';
  statusBar()->showMessage(buffer);
}

void Client_gpio::ledOn() {
  const char *cmd = "led_on";
  send_cmd( cmd );
}

void Client_gpio::ledOff() {
  const char *cmd = "led_off";
  send_cmd( cmd );
}

void Client_gpio::start_test() {
  const char *cmd = "start";
  send_cmd( cmd );
}

void Client_gpio::stop_test() {
  const char *cmd = "bye";
  send_cmd( cmd );
}

void Client_gpio::OnRed() {
  const char *cmd = "lcd_red";
  send_cmd( cmd );
}

void Client_gpio::OnGreen() {
  const char *cmd = "lcd_green";
  send_cmd( cmd );
}

void Client_gpio::OnBlue() {
  const char *cmd = "lcd_blue";
  send_cmd( cmd );
}

void Client_gpio::OnYellow() {
  const char *cmd = "lcd_yellow";
  send_cmd( cmd );
}

void Client_gpio::OnCyan() {
  const char *cmd = "lcd_cyan";
  send_cmd( cmd );
}

void Client_gpio::OnPurple() {
  const char *cmd = "lcd_purple";
  send_cmd( cmd );
}

void Client_gpio::OnWhite() {
  const char *cmd = "lcd_white";
  send_cmd( cmd );
}

void Client_gpio::OnBlack() {
  const char *cmd = "lcd_black";
  send_cmd( cmd );
}

void Client_gpio::setText(const QString &t) {
  m_text = t;
  update();
}

void Client_gpio::sendText() {
  char cmd[32];

  if (! m_text.isEmpty()) {
      QByteArray inBytes = m_text.toLocal8Bit();
      strcpy( cmd, "Hello" );
      strcat( cmd, inBytes.data() );
      send_cmd( cmd );
  }
}

void Client_gpio::servoRight() {
  const char *cmd = "servo_right";
  send_cmd( cmd );
}

void Client_gpio::servoLeft() {
  const char *cmd = "servo_left";
  send_cmd( cmd );
}

void Client_gpio::servoMiddle() {
  const char *cmd = "servo_middle";
  send_cmd( cmd );
}

void Client_gpio::servoStop() {
  const char *cmd = "servo_stop";
  send_cmd( cmd );
}

int main(int argc, char *argv[]) {

  QApplication app(argc, argv);

  Client_gpio window;

  window.resize(500, 600);
  window.setWindowTitle("gpio test");
  window.show();

  return app.exec();
}
